### ISSUE TITLE
question about 'sde' and 'cps' sampling when noise_level=0

### ISSUE THREAD

**Issue by MengHao666** [yifan123/flow_grpo#231]
when nosie level=0, I find the 'sde' sampling could get same result image with little difference compared with the official diffuser pipeline . But the 'cps' sampling method could get actually different result image with lots of difference.
Why?

**Comment by Jayce-Ping** [comment]
I'm interested in this question. Do you have any code to reproduce your results? I've tested the input and output of SDE and CPS with noise_level=0 and found they are identical, which should produce the same outputs. If you can provide the inference code, I would like to find out the root cause together!

**Comment by Jayce-Ping** [comment]
I've tested the outputs using [Flow-Factory](https://github.com/X-GenGroup/Flow-Factory/tree/main)'s scheduler:

```python
# Modified from https://huggingface.co/black-forest-labs/FLUX.1-dev usage example
import torch
from diffusers import FluxPipeline
from peft import PeftModel
from flow_factory.scheduler import FlowMatchEulerDiscreteSDEScheduler

with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
    # 1. Load pipeline
    pipe = FluxPipeline.from_pretrained(
        "black-forest-labs/FLUX.1-dev",
        torch_dtype=torch.bfloat16
    )
    pipe.enable_model_cpu_offload()
    scheduler_config = pipe.scheduler.config
    scheduler_sde = FlowMatchEulerDiscreteSDEScheduler(
        noise_level=0.0,
        dynamics_type='Flow-SDE',
        **scheduler_config
    )
    scheduler_cps = FlowMatchEulerDiscreteSDEScheduler(
        noise_level=0.0,
        dynamics_type='CPS',
        **scheduler_config
    )
    scheduler_ODE = FlowMatchEulerDiscreteSDEScheduler(
        noise_level=0.0,
        dynamics_type='ODE',
        **scheduler_config
    )

    # 2. Generate image with original scheduler
    prompt = "A cat holding a sign that says hello world"
    image_ori = pipe(
        prompt,
        height=1024,
        width=1024,
        guidance_scale=3.5,
        num_inference_steps=28,
        max_sequence_length=512,
        generator=torch.Generator("cpu").manual_seed(0)
    ).images[0]
    image_ori.save("flux-dev-ori.png")

    # 3. Generate image with SDE scheduler
    pipe.scheduler = scheduler_sde
    prompt = "A cat holding a sign that says hello world"
    image_sde = pipe(
        prompt,
        height=1024,
        width=1024,
        guidance_scale=3.5,
        num_inference_steps=28,
        max_sequence_length=512,
        generator=torch.Generator("cpu").manual_seed(0)
    ).images[0]
    image_sde.save("flux-dev-SDE.png")


    # 4. Generate image with CPS scheduler
    pipe.scheduler = scheduler_cps
    image_cps = pipe(
        prompt,
        height=1024,
        width=1024,
        guidance_scale=3.5,
        num_inference_steps=28,
        max_sequence_length=512,
        generator=torch.Generator("cpu").manual_seed(0)
    ).images[0]
    image_cps.save("flux-dev-CPS.png")


    # 5. Generate image with SDE scheduler using ODE
    pipe.scheduler = scheduler_ODE
    image_ODE = pipe(
        prompt,
        height=1024,
        width=1024,
        guidance_scale=3.5,
        num_inference_steps=28,
        max_sequence_length=512,
        generator=torch.Generator("cpu").manual_seed(0)
    ).images[0]
    image_ODE.save("flux-dev-ODE.png")


    # Compare images
    from PIL import Image
    import numpy as np
    def mse(image1, image2):
        arr1 = np.array(image1).astype(np.float32)
        arr2 = np.array(image2).astype(np.float32)
        return np.mean((arr1 - arr2) ** 2)

    print("MSE between original and SDE:", mse(image_ori, image_sde))
    print("MSE between original and CPS:", mse(image_ori, image_cps))
    print("MSE between original and ODE:", mse(image_ori, image_ODE))

```

```log
MSE between original and SDE: 11.455062
MSE between original and CPS: 6.0658174
MSE between original and ODE: 11.455062
```

It shows that:
1. CPS is the closest to the original diffusers' scheduler.
2. `SDE` and `ODE` are identitcal (further verified by mse), and sightly different from `CPS`.

Since `flow_grpo`'s repo and `Flow-Factory` repo both use `float32` in `scheduler.step`, I guess the root cause is numerial precision. Here are four images generated:


<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/c27447f5-26e2-4722-bb5d-a4d34fbedb5d" />

> flux-dev-ori.ong

<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/966cc064-7920-4655-ac93-e7fd75e17f80" />

> flux-dev-SDE.ong


<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/7ae36d07-217d-49a2-884e-a5fbf8cd53ea" />

> flux-dev-CPS.png

<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/b3c9f327-e68e-470d-98e0-22488fb411b6" />


> flux-dev-ODE.png

**Comment by MengHao666** [comment]
> I've tested the outputs using [Flow-Factory](https://github.com/X-GenGroup/Flow-Factory/tree/main)'s scheduler:
> 
> # Modified from https://huggingface.co/black-forest-labs/FLUX.1-dev usage example
> import torch
> from diffusers import FluxPipeline
> from peft import PeftModel
> from flow_factory.scheduler import FlowMatchEulerDiscreteSDEScheduler
> 
> with torch.autocast(device_type="cuda", dtype=torch.bfloat16):
>     # 1. Load pipeline
>     pipe = FluxPipeline.from_pretrained(
>         "black-forest-labs/FLUX.1-dev",
>         torch_dtype=torch.bfloat16
>     )
>     pipe.enable_model_cpu_offload()
>     scheduler_config = pipe.scheduler.config
>     scheduler_sde = FlowMatchEulerDiscreteSDEScheduler(
>         noise_level=0.0,
>         dynamics_type='Flow-SDE',
>         **scheduler_config
>     )
>     scheduler_cps = FlowMatchEulerDiscreteSDEScheduler(
>         noise_level=0.0,
>         dynamics_type='CPS',
>         **scheduler_config
>     )
>     scheduler_ODE = FlowMatchEulerDiscreteSDEScheduler(
>         noise_level=0.0,
>         dynamics_type='ODE',
>         **scheduler_config
>     )
> 
>     # 2. Generate image with original scheduler
>     prompt = "A cat holding a sign that says hello world"
>     image_ori = pipe(
>         prompt,
>         height=1024,
>         width=1024,
>         guidance_scale=3.5,
>         num_inference_steps=28,
>         max_sequence_length=512,
>         generator=torch.Generator("cpu").manual_seed(0)
>     ).images[0]
>     image_ori.save("flux-dev-ori.png")
> 
>     # 3. Generate image with SDE scheduler
>     pipe.scheduler = scheduler_sde
>     prompt = "A cat holding a sign that says hello world"
>     image_sde = pipe(
>         prompt,
>         height=1024,
>         width=1024,
>         guidance_scale=3.5,
>         num_inference_steps=28,
>         max_sequence_length=512,
>         generator=torch.Generator("cpu").manual_seed(0)
>     ).images[0]
>     image_sde.save("flux-dev-SDE.png")
> 
> 
>     # 4. Generate image with CPS scheduler
>     pipe.scheduler = scheduler_cps
>     image_cps = pipe(
>         prompt,
>         height=1024,
>         width=1024,
>         guidance_scale=3.5,
>         num_inference_steps=28,
>         max_sequence_length=512,
>         generator=torch.Generator("cpu").manual_seed(0)
>     ).images[0]
>     image_cps.save("flux-dev-CPS.png")
> 
> 
>     # 5. Generate image with SDE scheduler using ODE
>     pipe.scheduler = scheduler_ODE
>     image_ODE = pipe(
>         prompt,
>         height=1024,
>         width=1024,
>         guidance_scale=3.5,
>         num_inference_steps=28,
>         max_sequence_length=512,
>         generator=torch.Generator("cpu").manual_seed(0)
>     ).images[0]
>     image_ODE.save("flux-dev-ODE.png")
> 
> 
>     # Compare images
>     from PIL import Image
>     import numpy as np
>     def mse(image1, image2):
>         arr1 = np.array(image1).astype(np.float32)
>         arr2 = np.array(image2).astype(np.float32)
>         return np.mean((arr1 - arr2) ** 2)
> 
>     print("MSE between original and SDE:", mse(image_ori, image_sde))
>     print("MSE between original and CPS:", mse(image_ori, image_cps))
>     print("MSE between original and ODE:", mse(image_ori, image_ODE))
> ```
> MSE between original and SDE: 11.455062
> MSE between original and CPS: 6.0658174
> MSE between original and ODE: 11.455062
> ```
> 
> It shows that:
> 
> 1. CPS is the closest to the original diffusers' scheduler.
> 2. `SDE` and `ODE` are identitcal (further verified by mse), and sightly different from `CPS`.
> 
> Since `flow_grpo`'s repo and `Flow-Factory` repo both use `float32` in `scheduler.step`, I guess the root cause is numerial precision. Here are four images generated:
> 
> <img alt="Image" width="1024" height="1024" src="https://private-user-images.githubusercontent.com/93087577/553362049-c27447f5-26e2-4722-bb5d-a4d34fbedb5d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzE5NDAwMDgsIm5iZiI6MTc3MTkzOTcwOCwicGF0aCI6Ii85MzA4NzU3Ny81NTMzNjIwNDktYzI3NDQ3ZjUtMjZlMi00NzIyLWJiNWQtYTRkMzRmYmVkYjVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjI0VDEzMjgyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThlMjcxOTUwYzAwNzdjZTA5MzE3Yzc0MTdmYWFhNjc0ZDg4OWY4ZTZjMWE4Mzc5Yzg2NDg5MWQxZjQwMTMzMmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AcIeKYMyptZRbZNrhSz3MiNZ1uxNH2oUVExGxF1eN-I">
> > flux-dev-ori.ong
> 
> <img alt="Image" width="1024" height="1024" src="https://private-user-images.githubusercontent.com/93087577/553362123-966cc064-7920-4655-ac93-e7fd75e17f80.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzE5NDAwMDgsIm5iZiI6MTc3MTkzOTcwOCwicGF0aCI6Ii85MzA4NzU3Ny81NTMzNjIxMjMtOTY2Y2MwNjQtNzkyMC00NjU1LWFjOTMtZTdmZDc1ZTE3ZjgwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjI0VDEzMjgyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzNjFjNmJmYjZiNzRmYmE0Njg0Y2I1MWM1NzY3OTYxNjZmZDk1MTdiY2QwODE2OWNkY2QwYmZhN2UzYjRkMjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.S47h6JLwAIw0XpLf20beNcTVMeuH0D1_79zPgl6ep8c">
> > flux-dev-SDE.ong
> 
> <img alt="Image" width="1024" height="1024" src="https://private-user-images.githubusercontent.com/93087577/553362204-7ae36d07-217d-49a2-884e-a5fbf8cd53ea.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzE5NDAwMDgsIm5iZiI6MTc3MTkzOTcwOCwicGF0aCI6Ii85MzA4NzU3Ny81NTMzNjIyMDQtN2FlMzZkMDctMjE3ZC00OWEyLTg4NGUtYTVmYmY4Y2Q1M2VhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjI0VDEzMjgyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUxM2IxMTU0MzFiYTA2M2ZjODM3MTA5ZGE3NzNkOGZlNGQxYzY0NjRiZWQwMTVlZjQ2YTZkZTRmMWMwNzMxMGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RkIrKlDxaft4M9izdqJZEjs31c90rg_ifDaWH2SC3YA">
> > flux-dev-CPS.png
> 
> <img alt="Image" width="1024" height="1024" src="https://private-user-images.githubusercontent.com/93087577/553362229-b3c9f327-e68e-470d-98e0-22488fb411b6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzE5NDAwMDgsIm5iZiI6MTc3MTkzOTcwOCwicGF0aCI6Ii85MzA4NzU3Ny81NTMzNjIyMjktYjNjOWYzMjctZTY4ZS00NzBkLTk4ZTAtMjI0ODhmYjQxMWI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjI0VDEzMjgyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRlMzg4ZTUyZWZlZTI4MTY1MDI3YjE4YjdiMzBiNThlMjY4NWQyNDQzNDMxMzJlNDhiODUyOGFiODc3Nzg0NmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.F98cFJF7RDcUplJGVW0Lvzv7C2L-SmxOBVhyNTO3joc">
> > flux-dev-ODE.png

Thanks for codes and examples !  Considering your examples, I have found that the root cause was the data precision.
In my old implementaion, I dtype of  pipeline is `torch.bfloat16`, but the dtype of fixed input latents is "torch.float32", this may induce much difference between  `cps` sampling and `sde` sampling even when the noise level is zero.
